server/src/pages/shot/editor.js

```diff
@@ -0,0 +1,150 @@
+        </div>
+      </div>
+      <div className="main-container">
+        <div className="canvas-container" id="canvas-container" ref={(canvasContainer) => this.canvasContainer = canvasContainer}>
+          <canvas className="image-holder centered" id="image-holder" ref={(image) => { this.imageCanvas = image }} height={ 2 * canvasHeight } width={ 2 * canvasWidth } style={{height: canvasHeight, width: canvasWidth}}></canvas>
+          <canvas className="highlighter centered" id="highlighter" ref={(highlighter) => { this.highlighter = highlighter }} height={canvasHeight} width={canvasWidth}></canvas>
+          <canvas className="editor centered" id="editor" ref={(editor) => { this.editor = editor }} height={canvasHeight} width={canvasWidth}></canvas>
+        </div>
+      </div>
+    </div>
+  }
+
+  componentDidUpdate() {
+    this.edit();
+  }
+
+  onClickCancel() {
+    this.props.onCancelEdit(false);
+    sendEvent("cancel", "annotation-toolbar");
+  }
+
+  onClickSave() {
+    sendEvent("save", "annotation-toolbar");
+    this.imageContext.drawImage(this.editor, 0, 0);
+    this.imageContext.globalCompositeOperation = 'multiply';
+    this.imageContext.drawImage(this.highlighter, 0, 0);
+    let dataUrl = this.imageCanvas.toDataURL();
+    this.props.onClickSave(dataUrl);
+  }
+
+  onClickHighlight() {
+    if (this.state.tool != 'highlighter') {
+      this.setState({tool: 'highlighter'});
+      sendEvent("highlighter-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none'});
+      sendEvent("highlighter-deselect", "annotation-toolbar");
+    }
+  }
+
+  onClickPen() {
+    if (this.state.tool != 'pen') {
+      this.setState({tool: 'pen'});
+      sendEvent("pen-select", "annotation-toolbar");
+    } else {
+      this.setState({tool: 'none'});
+      sendEvent("pen-deselect", "annotation-toolbar");
+    }
+  }
+
+  componentDidMount() {
+    this.context = this.editor.getContext('2d');
+    this.highlightContext = this.highlighter.getContext('2d');
+    let imageContext = this.imageCanvas.getContext('2d');
+    this.imageContext = imageContext;
+    // From https://blog.headspin.com/?p=464, we oversample the canvas for improved image quality
+    let img = new Image();
+    img.src = this.props.clip.image.url;
+    let width = this.props.clip.image.dimensions.x;
+    let height = this.props.clip.image.dimensions.y;
+    this.imageContext.scale(2, 2);
+    this.imageContext.drawImage(img, 0, 0, width, height);
+  }
+
+  edit() {
+    pos = { x: 0, y: 0 };
+    if (this.state.tool == 'highlighter') {
+      this.highlightContext.lineWidth = 20;
+      this.highlightContext.strokeStyle = '#ff0';
+      this.drawContext = this.highlightContext;
+    } else if (this.state.tool == 'pen') {
+      this.context.strokeStyle = this.state.color;
+      this.context.globalCompositeOperation = 'source-over';
+      this.drawContext = this.context;
```

Comment:
I think the tool should be handled in `draw()` not here – so instead of swapping `this.drawContext` here, instead do it later.